### PR TITLE
Add [Agent] ServiceComb Plugin #437

### PR DESCRIPTION
**Pull request by lytscu**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [x] New feature provided

- Related issues
 #437
___

___
### New feature or improvement
- [Agent] ServiceComb Plugin

### CODE REVIEW

**Comment by wu-sheng** [comment]
@lytscu I just fixed your pr comments:
```
- [x] New feature provided
```

**Comment by wu-sheng** [comment]
@lytscu @ascrutae What is the status of this pr? now.

**Comment by wu-sheng** [comment]
@lytscu @ascrutae Are the codes ready? Does any test report to provide?

**Comment by ascrutae** [comment]
I found some question about this PR, here are the questions:

Q1:  I use the [demo](https://github.com/lytscu/skywalking-agent-testcases/tree/servicecomb/servicecomb-0.x-scenario) to trace the service about Servicecomb, I found the trace is broken,  The consumer trace segment should be like the following:
```
Span-0: codefirstClient.codeFirstSpringmvcHello.say
Span-1: codefirst.codeFirstHello.sayHi
```
but in fact those spans are in two segments.

Q2: The buried point `Invocation`  is more complex, I found some buried point about Servicecomb  plugin will more simple. 
**Consumer side**:
1. `io.servicecomb.common.rest.RestProducerInvocation#createInvocation` 
2. `io.servicecomb.common.rest.AbstractRestInvocation#invoke`

The `Pojo` way about Servicecomb
1. `io.servicecomb.provider.pojo.Invoker#invoke`

**Producer Side**:
1. `io.servicecomb.swagger.engine.SwaggerProducerOperation#invoke`

**Comment by wu-sheng** [comment]
@ascrutae @lytscu Why does the trace brake? Because of multi thread situation? or header propagation?

**Comment by wu-sheng** [comment]
@ascrutae I don't think these can be caused by where you set the interceptor.

**Comment by wu-sheng** [comment]
@ascrutae I can't follow you with the spans expected:
> Span-0: codefirstClient.codeFirstSpringmvcHello.say
Span-1: codefirst.codeFirstHello.sayHi

What does this mean? What is now? and brake how?

**Comment by ascrutae** [comment]
@wu-sheng Sorry for my poor english, and here is my explain.

I use the demo to enhance ServicceComb and found that the trace is broken. 
Here is the expected segment about demo:
```
 ServiceComb-Segment:
    Span-0: `codefirstClient.codeFirstSpringmvcHello.say`
    Span-1: `codefirst.codeFirstHello.sayHi`
```

In fact Here is the generated segment:

```
SerivceComb-Segment-0:
   Span-0: `codefirstClient.codeFirstSpringmvcHello.say`

ServiceComb-Segment-1:
  Span-0: `codefirst.codeFirstHello.sayHi`
```
    
The question is cause by the buired point is incorrect. Because of the enhance method `next` finished before the service method invoked.

I found some suggest enhance class about `REST` service. But the ServiceComb support multiple way to provide service, maybe @lytscu still find the similar enhance class. 

**Consumer side**
`io.servicecomb.common.rest.RestProducerInvocation#invoke`: The entry method about REST way to provide service.
`io.servicecomb.provider.pojo.Invoker#invoke`: The entry method about ServiceComb RPC client stub .(The demo in call ServiceComb service by PRC)

**Producer side**
`io.servicecomb.core.handler.impl.ProducerOperationHandler#handle`: The entry method about RPC service stub.

**Comment by wu-sheng** [comment]
@ascrutae That explain makes sense. 

@lytscu please pay attentiom. about you intercept point, make sure they are in the main request thread, such as http incoming request thread. same in server side.  And make sure header can propagate in the right way.

**Comment by wu-sheng** [comment]
@lytscu Do you finish the plugin? And make the test works?

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java [hunk 1]
```diff
@@ -48,16 +48,28 @@ public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allAr
...
             next.setHeadValue(invocation.getContext().get(next.getHeadKey()));
         }
+        if (null == invocation.getOperationMeta()) {
+            return;
+        }
         String operationName = invocation.getMicroserviceQualifiedName();
         AbstractSpan span = ContextManager.createEntrySpan(operationName, contextCarrier);
-        String url = invocation.getOperationMeta().getOperationPath();
-        Tags.URL.set(span, url);
+        if (null != invocation.getOperationMeta() && null != invocation.getOperationMeta().getOperationPath()) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java]: `(null == invocation.getOperationMeta()` has been invoked in line 51.

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java [hunk 2]
```diff
@@ -48,16 +48,28 @@ public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allAr
...
+            String url = invocation.getOperationMeta().getOperationPath();
+            Tags.URL.set(span, url);
+        }
         span.setComponent(ComponentsDefine.SERVICECOMB);
         SpanLayer.asRPCFramework(span);
     }
 
     @Override public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments,
         Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        Invocation invocation = (Invocation)allArguments[0];
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java]: If NO span  created, you can't guarantee active span exists.

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java]: And stop span can't be expected.

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java [hunk 3]
```diff
@@ -43,10 +43,14 @@
     public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments,
         Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
         Invocation invocation = (Invocation)allArguments[0];
+        if (null == invocation.getOperationMeta() || null == invocation.getEndpoint()) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java]: You should adjust stop span ,too.

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments,
+        Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+        Invocation invocation = (Invocation)allArguments[0];
+        ContextCarrier contextCarrier = new ContextCarrier();
+        CarrierItem next = contextCarrier.items();
+        while (next.hasNext()) {
+            next = next.next();
+            next.setHeadValue(invocation.getContext().get(next.getHeadKey()));
+        }
+        if (null == invocation.getOperationMeta()) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java]: When does Producer side face `null == invocation.getOperationMeta()`? Entry span requires nothing additional. You could create EntrySpan every time.

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        String url = invocation.getOperationMeta().getOperationPath();
+        Tags.URL.set(span, url);
+        span.setComponent(ComponentsDefine.SERVICECOMB);
+        SpanLayer.asRPCFramework(span);
+    }
+
+    @Override public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments,
+        Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        Invocation invocation = (Invocation)allArguments[0];
+        if (null == invocation.getOperationMeta()) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java]: When does Producer side face null == invocation.getOperationMeta()? Entry span requires nothing additional. You could create EntrySpan every time.

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ * @author lytscu
+ */
+public class TransportClientHandlerInterceptor implements InstanceMethodsAroundInterceptor {
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments,
+        Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+        Invocation invocation = (Invocation)allArguments[0];
+        if (null == invocation.getOperationMeta() || null == invocation.getEndpoint()) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java]: `if (null == invocation.getOperationMeta() || null == invocation.getEndpoint()) {` should be in a separated method.

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            Tags.STATUS_CODE.set(span, Integer.toString(statusCode));
+        }
+        ContextManager.stopSpan();
+
+        return ret;
+    }
+
+    @Override public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments,
+        Class<?>[] argumentsTypes, Throwable t) {
+        AbstractSpan span = ContextManager.activeSpan();
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java]: `#active` is not safe in exception handling, also need the `if (null == invocation.getOperationMeta() || null == invocation.getEndpoint()) {`

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        if (null == invocation.getOperationMeta() || null == invocation.getEndpoint()) {
+            return;
+        }
+        AbstractSpan span = ContextManager.activeSpan();
+        span.errorOccurred();
+        span.log(t);
+    }
+
+    /**
+     * Serviecomb chissis Consumers and providers need to register at the service center. If the consumer is not
```
Comment (ascrutae) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java]: `chissis` change to `chassis`

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.skywalking.apm.agent.core.context.tag.Tags;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+
+/**
+ * {org.apache.skywalking.apm.plugin.servicecomb.} define how to enhance class {@link Invocation#getHandlerContext()}.
```
Comment (ascrutae) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java]: The comments is incorrect, please adjust it

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments,
+        Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+        Invocation invocation = (Invocation)allArguments[0];
+        ContextCarrier contextCarrier = new ContextCarrier();
+        CarrierItem next = contextCarrier.items();
+        while (next.hasNext()) {
+            next = next.next();
+            next.setHeadValue(invocation.getContext().get(next.getHeadKey()));
+        }
+        String operationName = "servicecomb/chassis" + method.getName();
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/ProducerOperationHandlerInterceptor.java]: Default name should be `className` + `methodName`, and only execute in `else`.

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+     * registered then return
+     */
+    private String AsserRegister(Invocation invocation) throws URISyntaxException {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/TransportClientHandlerInterceptor.java]: You misunderstand what I mean about this method. This method should be `#checkRegisterStatus` and return true/false. The method will make the codes more clear.

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/define/TransportClientHandlerInstrumentation.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.ConstructorInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.InstanceMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+import org.apache.skywalking.apm.agent.core.plugin.match.NameMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+/**
+ * {@link TransportClientHandlerInstrumentation} presents that skywalking intercept {@link TransportClientHandler}by
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/define/TransportClientHandlerInstrumentation.java]: `presents` to `represents`

File: apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/define/ProducerOperationHandlerInstrumentation.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.ConstructorInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.InstanceMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+import org.apache.skywalking.apm.agent.core.plugin.match.NameMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+/**
+ * {@link ProducerOperationHandlerInstrumentation} presents that skywalking intercept {@link
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/servicecomb-plugin/servicecomb-java-chassis-0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/servicecomb/define/ProducerOperationHandlerInstrumentation.java]: `represents` to `represents`

**Comment by ascrutae** [comment]
Here is the test report: https://github.com/SkywalkingTest/agent-integration-test-report/blob/master/2018/2/lytscu/testReport-master-2018-02-12-15-28.md

**Comment by lytscu** [comment]
https://github.com/SkywalkingTest/agent-integration-test-report/blob/master/2018/2/lytscu/testReport-master-2018-02-12-16-49.md

**Comment by wu-sheng** [comment]
@lytscu @ascrutae Seem like, ServiceComb 0.1.0 works out magically.

**Comment by wu-sheng** [comment]
@lytscu Please add the `serviceComb` to supported list.

**Comment by wu-sheng** [comment]
FYI @WillemJiang , the plugin is checked by our automatic mechanism and merged.

**Pull request merged**